### ISSUE TITLE
Segfault and incorrect results for simple aggregation

### ISSUE THREAD

**Issue by eric-czech** [hail-is/hail#8076]
I found that aggregating sorted arrays gives incorrect results and crashes when trying to do so without converting the array to string.  This code explains what I found pretty well:

```python
> mt = hl.balding_nichols_model(1, 10, 10)

# Aggregate concatenated alleles (works fine)
> mt.aggregate_rows(hl.agg.counter(hl.delimit(mt.alleles, '|')))
{'A|C': 10}

# Group by the array directly (gives an expected error)
> mt.aggregate_rows(hl.agg.counter(mt.alleles))
TypeError: unhashable type: 'list'

# Aggregate sorted arrays (works but gives wrong result)
> mt.aggregate_rows(hl.agg.counter(hl.delimit(hl.sorted(mt.alleles), '|')))
{'A|A|A|C|\x0b\x00\x00': 2, 'A|A|A|C|C|C': 8}

# Aggregate the sorted arrays directly (segfault)
# *This should probably throw "unhashable type list" like it does without the sort*
mt.aggregate_rows(hl.agg.counter(hl.sorted(mt.alleles)))
ERROR:root:Exception while sending command.
Traceback (most recent call last):
  File "/opt/conda/envs/hail/lib/python3.7/site-packages/py4j/java_gateway.py", line 1159, in send_command
    raise Py4JNetworkError("Answer from Java side is empty")
py4j.protocol.Py4JNetworkError: Answer from Java side is empty
...
Py4JError: An error occurred while calling o59.executeJSON
```

Here is the full [stack trace](https://github.com/hail-is/hail/files/4187400/stacktrace.txt) and [core dump](https://github.com/hail-is/hail/files/4187399/coredump.txt).

I think some related questions that arise from this are:

1. What's the best way to group by an array to avoid the conversion to a delimited string?  In this case I could do something like ```mt.aggregate_rows(hl.agg.counter(hl.tuple([mt.alleles[0], mt.alleles[1]])))``` but I can't find a solution for getting a tuple from an array without knowing the length of it beforehand for every row.  Is there a more fundamental reason why the API doesn't allow aggregation by arrays even if Spark does?
2. When the Py4J server crashes, it's no longer reachable from the python clients so I have to restart my process and re-initialize Hail.  Is there already functionality implemented for bringing that server up if it's down?  I'd imagine segfaults aren't the only reason it could down, so it would be nice if there was a way to bring it back up either automatically or manually.

Hail version: 0.2.30-2ae07d872f43
Spark version: 2.4.4

**Comment by tpoterba** [comment]
oof, good find. Thanks for the bug report, we'll fix ASAP.

**Comment by tpoterba** [comment]
randomly assigned @catoverdrive

**Comment by tpoterba** [comment]
to answer (1), I think the right thing is for us to implement our own hashable immutable data structures (and use frozenset for sets, for instance) for results of Hail computations. I think we have yet to nail down whether this would be a breaking interface change, forcing us to wait until 0.3.

To answer (2), you *may* be able to do `hl.stop(); hl.init()` to reset the session, but not sure this will work in every case. The driver should really only die for OOM and faults, I think.

**Comment by catoverdrive** [comment]
Hi @eric-czech,

The errors you're running into with aggregating sorted arrays should be fixed in versions 0.2.31 and later; if upgrading the version works for you, please let me know and I will close this issue.

**Comment by eric-czech** [comment]
Thanks @catoverdrive!  I pulled 0.2.32 and got the expected results for all of those.
